### PR TITLE
BFB-288 / BFB-308 - Melee Animation & Damage Offset

### PR DESCRIPTION
**Pull request by unoctium1**
**JIRA Ticket**
https://perigongames.atlassian.net/browse/BFB-288 / https://perigongames.atlassian.net/browse/BFB-308

**Description**
Add placeholder animation to melee attack and and offset damage

Note that this kinda devolved into a bit of a callback hell situation - animation is associated with the WeaponManager, due to the scene hierarchy there, but melee is all handled by the equipment manager. Probably should wait to merge until our meeting on Saturday to rearchitect the equipment

Edit:

Includes a fairly heavy refactor of equipment behavior and weapons, weapons now no longer have to be under the equipment manager, no longer need their own bullet spawner, and handle their own animations. Each weapon behavior handles its own model, as well. Also converted the animation callback script into a generic script in the utility folder.

### CODE REVIEW

File: Assets/_BForBoss/_Weapons/Scripts/Equipment/EquipmentBehaviour.cs [hunk 1]
```diff
@@ -20,12 +20,19 @@ public partial class EquipmentBehaviour : MonoBehaviour
         private InputAction _swapWeaponInputAction = null;
         private InputAction _meleeWeaponInputAction = null;
         private bool _isMouseScrollEnabled = true;
+
+        public Action MeleeAttack;
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Weapons/Scripts/Equipment/EquipmentBehaviour.cs:24]: public should be above private variables

File: Assets/_BForBoss/_Weapons/Scripts/Equipment/WeaponAnimationCallbackHandler.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Weapons/Scripts/Equipment/WeaponAnimationCallbackHandler.cs:3]: Is this directory needed?

File: Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs [hunk 3]
```diff
@@ -27,33 +28,36 @@ public void DecrementCooldown(float deltaTime)
             _currentCooldown -= _currentCooldown > 0 ? deltaTime : _currentCooldown;
         }
 
-        public void AttackManyIfPossible(Vector3 playerPosition, Vector3 playerForwardDirection)
+        public bool AttackManyIfPossible(Vector3 playerPosition, Vector3 playerForwardDirection)
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs:31]: since this now returns a `bool` would `TryAttackMany` be a more suitable name

File: Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs [hunk 4]
```diff
@@ -27,33 +28,36 @@ public void DecrementCooldown(float deltaTime)
...
-            for(int i = 0; i < hits; i++)
-            {
-                DamageEnemy(_enemyBuffer[i]);
-            }
+            _hits = _meleeProperties.OverlapCapsule(playerPosition, playerForwardDirection, ref _enemyBuffer);
+            return true;
         }
         
-        public void AttackOneIfPossible(Vector3 playerPosition, Vector3 playerForwardDirection)
+        public bool AttackOneIfPossible(Vector3 playerPosition, Vector3 playerForwardDirection)
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs:41]: Similarly `CanAttackOne`

File: Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs [hunk 5]
```diff
@@ -27,33 +28,36 @@ public void DecrementCooldown(float deltaTime)
...
         {
             if (!CanMelee)
-                return;
+                return false;
             _currentCooldown += _meleeProperties.AttackCoolDown;
             
             var hits = _meleeProperties.OverlapCapsule(playerPosition, playerForwardDirection, ref _enemyBuffer);
-            if (hits <= 0) 
+            if (hits > 1)
+                hits = 1; //ensure we only damage first enemy
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs:49]: should this be `_hits` or am I missing something?

File: Assets/_BForBoss/_Weapons/Scripts/Weapons/WeaponsManager.cs [hunk 6]
```diff
@@ -11,6 +11,9 @@ namespace Perigon.Weapons
     public class WeaponsManager : MonoBehaviour
     {
         private const float ACCUMULATED_RECOIL_PERCENTAGE = 0.99F;
+        private const string MELEE_ANIMATOR_PARAM = "Melee";
+        private static int MeleeAnimatorID = Animator.StringToHash(MELEE_ANIMATOR_PARAM);
+
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Weapons/Scripts/Weapons/WeaponsManager.cs:16]: Up to you, as it is pretty unimportant now, but since we will be having enemies and bosses with their own animation, it might be better to have `MELEE_ANIMATOR_PARAM` string be something like `Melee_Player` or some thing like that!

File: Assets/_BForBoss/_Weapons/Scripts/Weapons/WeaponsManager.cs [hunk 7]
```diff
@@ -63,9 +68,15 @@ public class WeaponsManager : MonoBehaviour
...
         }
+        
+        public void MeleeHitCallback()
+        {
+            _equipmentBehaviour.ApplyMeleeDamageDelayed();
+        }
 
         private void Start()
         {
+            _weaponAnimator = GetComponentInChildren<Animator>();
```
Comment (Gursaanj) [inline Assets/_BForBoss/_Weapons/Scripts/Weapons/WeaponsManager.cs:79]: Up to you, but does this have to be on `Start()` or can we just attach an Animator to `_weaponAnimator` via `[SerializeField]`

**Review by Gursaanj — COMMENTED** [review]
Looks good, Mostly some nit-picky comments

File: Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs (hunk 5 above)
Comment (unoctium1) [reply Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs:49]: You're totally right, good catch!

**Comment by m4rtint** [comment]
image
Is this setup correctly?

File: Assets/_BForBoss/_Core/Scripts/Managers/GunRangeWorldManager.cs [hunk 8]
```diff
@@ -4,13 +4,15 @@
 using Perigon.Weapons;
 using Sirenix.OdinInspector;
 using UnityEngine;
+using UnityEngine.Serialization;
 
 namespace BForBoss
 {
     public class GunRangeWorldManager : BaseWorldManager
     {
+        [FormerlySerializedAs("_weaponsManager")]
```
Comment (m4rtint) [inline Assets/_BForBoss/_Core/Scripts/Managers/GunRangeWorldManager.cs:13]: You can probably remove this?

File: Assets/_BForBoss/_Core/Scripts/Managers/GunRangeWorldManager.cs [hunk 9]
```diff
@@ -45,7 +47,7 @@ protected override void Reset()
         protected override void OnValidate()
         {
             base.OnValidate();
-            if (_weaponsManager == null)
+            if (weaponAnimationController == null)
             {
                 PanicHelper.Panic(new Exception("Weapons Manager missing from World Manager"));
```
Comment (m4rtint) [inline Assets/_BForBoss/_Core/Scripts/Managers/GunRangeWorldManager.cs:52]: Can you reword this as well

File: Assets/_BForBoss/_Core/Scripts/Managers/WeaponRaceCourseWorldManager.cs [hunk 10]
```diff
@@ -107,7 +109,7 @@ protected override void HandleOnDeath()
         protected override void OnValidate()
         {
             base.OnValidate();
-            if (_weaponsManager == null)
+            if (weaponAnimationController == null)
             {
                 PanicHelper.Panic(new Exception("Weapons Manager missing from World Manager"));
```
Comment (m4rtint) [inline Assets/_BForBoss/_Core/Scripts/Managers/WeaponRaceCourseWorldManager.cs:114]: Same here, reword this to match

File: Assets/_BForBoss/_Core/Scripts/Managers/WeaponSandboxWorldManager.cs [hunk 11]
```diff
@@ -20,8 +21,9 @@ public class WeaponSandboxWorldManager : BaseWorldManager
         [Title("Component")]
         [SerializeField] private LifeCycleManager _lifeCycleManager = null;
         
+        [FormerlySerializedAs("_weaponsManager")]
```
Comment (m4rtint) [inline Assets/_BForBoss/_Core/Scripts/Managers/WeaponSandboxWorldManager.cs:24]: Can remove this?

File: Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeaponBehaviour.cs [hunk 12]
```diff
@@ -42,10 +47,14 @@ private void OnMeleeInputAction(InputAction.CallbackContext context)
             if (context.performed)
             {
                 var t = _playerTransform ? _playerTransform : _getTransform();
-                if(_canAttackMany)
-                    _weapon.AttackManyIfPossible(t.position, t.forward);
-                else
-                    _weapon.AttackOneIfPossible(t.position, t.forward);
+                var attack = _canAttackMany ?
```
Comment (m4rtint) [inline Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeaponBehaviour.cs:50]: can you rename this to.. `isAttackSuccesful`

File: Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs [hunk 13]
```diff
@@ -27,33 +28,36 @@ public void DecrementCooldown(float deltaTime)
...
-            if (hits <= 0) 
+            _hits = _meleeProperties.OverlapCapsule(playerPosition, playerForwardDirection, ref _enemyBuffer);
+            if (_hits > 1)
+                _hits = 1; //ensure we only damage first enemy
+            return true;
+        }
+        
+        public void ApplyDamageDelayed()
+        {
+            if (_enemyBuffer == null)
```
Comment (m4rtint) [inline Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs:55]: _enemyBugger.IsEmptyOrNull()

use this instaed?

Comment (m4rtint) [reply Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs:55]: Why is it called Apply Damaged Delayed?

Comment (unoctium1) [reply Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs]: That was a holdover from when delaying damage was a toggle, I'll rename it

File: Assets/_BForBoss/_Core/Scripts/Managers/GunRangeWorldManager.cs [hunk 14]
```diff
@@ -4,13 +4,14 @@
...
 using UnityEngine;
+using UnityEngine.Serialization;
 
 namespace BForBoss
 {
     public class GunRangeWorldManager : BaseWorldManager
     {
         [Title("Component")] 
-        [SerializeField] private WeaponsManager _weaponsManager = null;
+        [SerializeField] private WeaponAnimationController weaponAnimationController = null;
```
Comment (m4rtint) [inline Assets/_BForBoss/_Core/Scripts/Managers/GunRangeWorldManager.cs:14]: I'll approve, this should be `_weaponAnimationController`. 
same with the other world manager

**Review by m4rtint — APPROVED** [review]

File: Assets/_BForBoss/_Core/Scripts/Managers/GunRangeWorldManager.cs (hunk 14 above)
Comment (unoctium1) [reply Assets/_BForBoss/_Core/Scripts/Managers/GunRangeWorldManager.cs:14]: Ah thats the rider rafactor screwing up here haha, I'll fix that and then merge

File: Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs [hunk 15]
```diff
@@ -14,7 +15,8 @@ public class MeleeWeapon
         private float _currentCooldown = 0f;
         private Action<bool> _onHitEntity;
         public bool CanMelee => _currentCooldown <= 0f;
-
+        private int _hits;
+        
         public float CurrentCooldown => _currentCooldown;
 
         public MeleeWeapon(IMeleeProperties meleeProperties, Action<bool> onHitEntity = null)
```
[the remark this answers is not in the document]
Comment (unoctium1) [reply Assets/_BForBoss/_Weapons/Scripts/Weapons/MeleeWeapon.cs:22]: Yeah, thats intentional - I was thinking it would be handy for UI or vfx, similar to the onHitEntity callback on guns
